### DISCUSSION TITLE
Ideas for improvements to `Stream` trait

### DISCUSSION THREAD

**Discussion by Mesoptier** [winnow-rs/winnow#519] [Ideas]
I'm working on a project where I'd like to implement a `Stream` that produces tokens just-in-time and discards its incoming data when it is no longer needed. As far as I can tell this is not current possible with the `Stream` traits as they are.

### `eof_offset`
In particular, some of the offset related methods make no sense when producing tokens just-in-time. I don't think there's anything sensible I can return from `.eof_offset()` for example. Searching the codebase for that method, it looks like the main uses are:
1. To see if the stream has reached its end. → This could simply be a `Stream::is_eof(&self) -> bool` method, which would be easier to implement for different stream types. 
   Couldn't it also just be based on whether `next_token()` returns `None`? I guess maybe not since this could also happen when more data is needed, though maybe it should return a `Result<Token, IncompleteOrEOF>` then?
2. To see if a parser consumed anything. → Could be replaced with a `Stream::consumed_since_checkpoint(&self, Self::Checkpoint) -> bool` method.
3. To actually get the offset to the end. → For these cases a separate `EofOffset` trait could be added. Or maybe something more generic to indicate "this stream knows when it ends".

### Checkpoints

The only feedback a stream gets regarding checkpoints is when the stream is reset to one. It would be nice if the interface was set up such that the stream can also tell when a checkpoint is dropped, so it can in turn drop any data from before the checkpoint. 

Maybe I should just use `std::ops::Drop` for this, but it feels to me like an interface for this on Stream would be nice, no?

Additionally, I don't think `Stream::Checkpoint` should have `Clone`. If a combinator needs more checkpoints, it should just call `.checkpoint()` multiple times.

**Comment by yyy33** [comment]
In order not to make too many changes, maybe we can split Stream and keep only its core functionality. For example
````rust
pub trait EofOffsetStream: Stream {
fn eof_offset(&self);
}
````

**Reply by yyy33** [reply]
Can you provide more specific examples?

**Reply by Mesoptier** [reply]
I haven't worked on the project I mentioned in the OP for a while, so I'm not "in it" enough to provide any examples at the moment, sorry.

**Comment by epage** [comment]
Building up a list of related discussions
- #528
- #232
- #299
- #572

**Reply by epage** [reply]
From #528 

> nom v7 had more individual traits, letting parsers declare what their requirements are.
>
> This made it more difficult to make custom streams, more difficult to make generic parsers, and makes the parsers brittle where subtle implementation changes can change the requirements on the traits. We resolved this by combining almost all of the traits into Stream.
>
> Even if we split out contiguous and token-focused streams, I wonder if there would be enough functionality left to be of us. If there is enough of a case for token-only parsing to be sufficient, we can revisit the trait design.

**Reply by chifflier** [reply]
Hi,

Sorry if this looks like hijacking the topic, but since the discussion is about the streaming API I would like to give some feedback.

nom allows to combine easily streaming and non-streaming parsers easily (and some combinators to help, like `complete`). This is not perfect, but it makes combining, for examples, librairies easily.
The problem I see in current `winnow` design is that, since streaming parsers have to use a different type (`Partial`), combining different parsers is not easy anymore. Also, if not using the streaming API, it is forbidden to use `Incomplete(n)` as return.

Workarounds include wrapping calls (quite expensive) or rewriting everything to use a generic type as Input, which creates other problems (the `Partial` is included in the error type as well, so raising an error still prevents combining functions). I tried to explain that in #507
